fmt/src/visit.rs

```diff
@@ -0,0 +1,268 @@
+        Ok(())
+    }
+
+    fn visit_var_def(&mut self, var: &mut VariableDefinition) -> VResult {
+        if !var.doc.is_empty() {
+            self.visit_doc_comments(&mut var.doc)?;
+            self.visit_newline()?;
+        }
+        self.visit_source(var.loc)?;
+        self.visit_stray_semicolon()?;
+
+        Ok(())
+    }
+
+    fn visit_return(&mut self, _expr: &mut Option<Expression>) -> VResult {
+        Ok(())
+    }
+
+    fn visit_break(&mut self) -> VResult {
+        Ok(())
+    }
+
+    fn visit_continue(&mut self) -> VResult {
+        Ok(())
+    }
+
+    fn visit_do_while(&mut self, _stmt: &mut Statement, _expr: &mut Expression) -> VResult {
+        Ok(())
+    }
+
+    fn visit_while(&mut self, _expr: &mut Expression, _stmt: &mut Statement) -> VResult {
+        Ok(())
+    }
+
+    fn visit_function(&mut self, func: &mut FunctionDefinition) -> VResult {
+        if !func.doc.is_empty() {
+            self.visit_doc_comments(&mut func.doc)?;
+            self.visit_newline()?;
+        }
+        self.visit_source(func.loc())?;
+        if func.body.is_none() {
+            self.visit_stray_semicolon()?;
+        }
+
+        Ok(())
+    }
+
+    fn visit_struct(&mut self, structure: &mut StructDefinition) -> VResult {
+        if !structure.doc.is_empty() {
+            self.visit_doc_comments(&mut structure.doc)?;
+            self.visit_newline()?;
+        }
+        self.visit_source(structure.loc)?;
+
+        Ok(())
+    }
+
+    fn visit_event(&mut self, event: &mut EventDefinition) -> VResult {
+        if !event.doc.is_empty() {
+            self.visit_doc_comments(&mut event.doc)?;
+            self.visit_newline()?;
+        }
+        self.visit_source(event.loc)?;
+        self.visit_stray_semicolon()?;
+
+        Ok(())
+    }
+
+    fn visit_error(&mut self, error: &mut ErrorDefinition) -> VResult {
+        if !error.doc.is_empty() {
+            self.visit_doc_comments(&mut error.doc)?;
+            self.visit_newline()?;
+        }
+        self.visit_source(error.loc)?;
+        self.visit_stray_semicolon()?;
+
+        Ok(())
+    }
+
+    fn visit_event_parameter(&mut self, param: &mut EventParameter) -> VResult {
+        self.visit_source(param.loc)?;
+
+        Ok(())
+    }
+
+    fn visit_stray_semicolon(&mut self) -> VResult;
+
+    fn visit_newline(&mut self) -> VResult;
+
+    fn visit_using(&mut self, using: &mut Using) -> VResult {
+        self.visit_source(using.loc)?;
+        self.visit_stray_semicolon()?;
+
+        Ok(())
+    }
+}
+
+/// All `solang::pt::*` types, such as [Statement](solang::pt::Statement) should implement the
+/// [Visitable] trait that accepts a trait [Visitor] implementation, which has various callback
+/// handles for Solidity Parse Tree nodes.
+///
+/// We want to take a `&mut self` to be able to implement some advanced features in the future such
+/// as modifying the Parse Tree before formatting it.
```

Comment:
Modifying the tree meaning e.g. preprocessor?

Reply:
more like having something like `forge fix` in the future, as Matt said:
> The only nit I’d have is keeping everything mut in Visitable and Visitor
This would allow us to implement something like forge fix in the future, where we modify the AST while traversing it